### PR TITLE
Move embedding to phi

### PR DESCRIPTION
**Pull request by phlrain**
### PR types
Breaking changes

### PR changes
OPs

### Describe

move embedding to phi

### CODE REVIEW

File: paddle/phi/kernels/cpu/embedding_grad_kernel.cc [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+#include "paddle/fluid/framework/convert_utils.h"
+#include "paddle/fluid/framework/data_type.h"
+#include "paddle/phi/backends/cpu/cpu_context.h"
+#include "paddle/phi/core/kernel_registry.h"
+
+namespace phi {
+
+template <typename T, typename Context>
+struct LookupTableV2GradCPUFunctor {
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/embedding_grad_kernel.cc]: 在phi下面LookupTableV2要不要都统一成Embedding？

File: paddle/phi/kernels/cpu/embedding_grad_kernel.cc [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    DDim table_dim = weight_.dims();
+
+    auto ids = CopyIdsToVector<IdT, int64_t>(input_);
+    auto ids_num = static_cast<int64_t>(ids.size());
+
+    // Since paddings are not trainable and fixed in forward, the gradient of
+    // paddings makes no sense and we don't deal with it in backward.
+    {
+      auto* d_output = &out_grad_;
+      // auto d_table = weight_grad_;
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/embedding_grad_kernel.cc]: 注释可以删除

File: paddle/phi/kernels/cpu/embedding_grad_kernel.cc [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+      for (int64_t i = 0; i < ids_num; ++i) {
+        if (padding_idx_ != kNoPadding && ids_data[i] == padding_idx_) {
+          // the gradient of padding_idx should be 0, already done by memset, so
+          // do nothing.
+        } else {
+          PADDLE_ENFORCE_LT(
+              ids_data[i],
+              N,
+              phi::errors::InvalidArgument(
+                  "Variable value (input) of OP(fluid.layers.embedding) "
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/embedding_grad_kernel.cc]: fluid.layers.embedding->paddle.nn.functional.embedding

File: paddle/phi/kernels/cpu/embedding_grad_kernel.cc [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                  "expected >= 0 and < %ld, but got %ld. Please check input "
+                  "value.",
+                  N,
+                  ids_data[i]));
+          PADDLE_ENFORCE_GE(
+              ids_data[i],
+              0,
+              phi::errors::InvalidArgument(
+                  "Variable value (input) of OP(fluid.layers.embedding) "
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/embedding_grad_kernel.cc]: fluid.layers.embedding->paddle.nn.functional.embedding

File: paddle/phi/kernels/cpu/embedding_grad_kernel.cc [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  const Context& dev_ctx_;
+  const DenseTensor& input_;
+  const DenseTensor& weight_;
+  const DenseTensor& out_grad_;
+  SelectedRows* weight_grad_;
+  int64_t padding_idx_;
+};
+
+template <typename T, typename Context>
+void EmbeddingSparseGradKernel(const Context& ctx,
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/embedding_grad_kernel.cc:186]: 这个Kernel可以放到selected_rows下

File: paddle/phi/kernels/cpu/embedding_grad_kernel.cc [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/phi/kernels/embedding_grad_kernel.h"
+#include "paddle/phi/kernels/funcs/embedding_util.h"
+
+#include "paddle/fluid/framework/convert_utils.h"
+#include "paddle/fluid/framework/data_type.h"
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/embedding_grad_kernel.cc]: 使用phi下的data_type.h

File: paddle/phi/kernels/cpu/embedding_kernel.cc [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+    for (int64_t i = 0; i < ids_numel; ++i) {
+      if (padding_idx_ != kNoPadding && ids[i] == padding_idx_) {
+        memset(output + i * row_width, 0, row_width * sizeof(T));
+      } else {
+        PADDLE_ENFORCE_LT(
+            ids[i],
+            row_number,
+            phi::errors::InvalidArgument(
+                "Variable value (input) of OP(fluid.layers.embedding) "
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/embedding_kernel.cc:59]: fluid.layers.embedding->paddle.nn.functional.embedding

File: paddle/phi/kernels/cpu/embedding_kernel.cc [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+                "expected >= 0 and < %ld, but got %ld. Please check input "
+                "value.",
+                row_number,
+                ids[i]));
+        PADDLE_ENFORCE_GE(
+            ids[i],
+            0,
+            phi::errors::InvalidArgument(
+                "Variable value (input) of OP(fluid.layers.embedding) "
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/embedding_kernel.cc:68]: fluid.layers.embedding->paddle.nn.functional.embedding

File: paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+#include "paddle/phi/kernels/funcs/embedding_util.h"
+
+#include "paddle/fluid/framework/convert_utils.h"
+#include "paddle/fluid/framework/data_type.h"
+#include "paddle/phi/backends/cpu/cpu_context.h"
+#include "paddle/phi/core/kernel_registry.h"
+
+namespace phi {
+
+template <typename T, typename Context>
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc:25]: 这个文件可以放到selected_rows目录下

File: paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/phi/kernels/sparse_weight_embedding_grad_kernel.h"
+#include "paddle/phi/kernels/funcs/embedding_util.h"
+
+#include "paddle/fluid/framework/convert_utils.h"
+#include "paddle/fluid/framework/data_type.h"
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc]: 引用phi下的data_type.h

File: paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+                  "Variable value (input) of OP(fluid.layers.embedding) "
+                  "expected >= 0 and < %ld, but got %ld. Please check input "
+                  "value.",
+                  N,
+                  ids_data[i]));
+          PADDLE_ENFORCE_GE(
+              ids_data[i],
+              0,
+              phi::errors::InvalidArgument(
+                  "Variable value (input) of OP(fluid.layers.embedding) "
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc]: fluid.layers.embedding->paddle.nn.functional.embedding

File: paddle/phi/kernels/cpu/sparse_weight_embedding_kernel.cc [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+  template <typename IdT>
+  void apply() {
+    auto ids = CopyIdsToVector<IdT, int64_t>(input_);
+    auto ids_numel = static_cast<int64_t>(ids.size());
+
+    const auto& table_t = weight_;
+    auto output_t = out_;
+    int64_t row_width = table_t.value().dims()[1];
+    const auto* table = table_t.value().template data<T>();
+    auto* output = output_t->template mutable_data<T>(dev_ctx_.GetPlace());
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/sparse_weight_embedding_kernel.cc]: 使用dec_ctx.Alloc<T>

File: paddle/phi/kernels/cpu/sparse_weight_embedding_kernel.cc [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    auto input_data_type =
+        paddle::framework::TransToProtoVarType(table_t.value().dtype());
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/sparse_weight_embedding_kernel.cc:50]: 这里可以不转ProtoVarType，直接用dtype

**Review by zyfncg — DISMISSED** [review]

File: paddle/phi/kernels/cpu/sparse_weight_embedding_kernel.cc (hunk 13 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/sparse_weight_embedding_kernel.cc:50]: done

File: paddle/phi/kernels/cpu/sparse_weight_embedding_kernel.cc (hunk 12 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/sparse_weight_embedding_kernel.cc]: done

File: paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc (hunk 11 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc]: done

File: paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc (hunk 10 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc]: done

File: paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc (hunk 9 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc:25]: selected rows的拆分，单独用一个pr来做

File: paddle/phi/kernels/cpu/embedding_kernel.cc (hunk 8 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/embedding_kernel.cc:68]: done

File: paddle/phi/kernels/cpu/embedding_kernel.cc (hunk 7 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/embedding_kernel.cc:59]: done

File: paddle/phi/kernels/cpu/embedding_grad_kernel.cc (hunk 6 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/embedding_grad_kernel.cc]: done

File: paddle/phi/kernels/cpu/embedding_grad_kernel.cc (hunk 5 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/embedding_grad_kernel.cc:186]: selected rows的拆分，单独用一个pr来做

File: paddle/phi/kernels/cpu/embedding_grad_kernel.cc (hunk 4 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/embedding_grad_kernel.cc]: done

File: paddle/phi/kernels/cpu/embedding_grad_kernel.cc (hunk 3 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/embedding_grad_kernel.cc]: done

File: paddle/phi/kernels/cpu/embedding_grad_kernel.cc (hunk 2 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/embedding_grad_kernel.cc]: done

File: paddle/phi/kernels/cpu/embedding_grad_kernel.cc (hunk 1 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/embedding_grad_kernel.cc]: done

File: paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+                                     DenseTensor* weight_grad) {
+  SparseWeightEmbeddingGradCPUFunctor<T, Context> functor(
+      ctx, input, weight, out_grad, padding_idx, weight_grad);
+
+  if (input.dtype() == phi::DataType::INT32) {
+    functor.template apply<int>();
+  } else if (input.dtype() == phi::DataType::INT64) {
+    functor.template apply<int64_t>();
+  } else {
+    PADDLE_THROW("emebdding input only support int32 and int64");
```
Comment (chenwhql) [inline paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc]: 这不符合报错规范，需要加报错类型

File: paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+                                           SelectedRows* weight_grad) {
+  SparseWeightEmbeddingSparseGradCPUFunctor<T, Context> functor(
+      ctx, input, weight, out_grad, padding_idx, weight_grad);
+
+  if (input.dtype() == phi::DataType::INT32) {
+    functor.template apply<int>();
+  } else if (input.dtype() == phi::DataType::INT64) {
+    functor.template apply<int64_t>();
+  } else {
+    PADDLE_THROW("emebdding input only support int32 and int64");
```
Comment (chenwhql) [inline paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc]: 同上

File: paddle/phi/kernels/gpu/embedding_grad_kernel.cu [hunk 16]
```diff
@@ -0,0 +1,256 @@
...
+                         DenseTensor* weight_grad) {
+  EmbeddingGradCUDAFunctor<T, Context> functor(
+      ctx, input, weight, out_grad, padding_idx, weight_grad);
+
+  if (input.dtype() == phi::DataType::INT32) {
+    functor.template apply<int>();
+  } else if (input.dtype() == phi::DataType::INT64) {
+    functor.template apply<int64_t>();
+  } else {
+    PADDLE_THROW("emebdding input only support int32 and int64");
```
Comment (chenwhql) [inline paddle/phi/kernels/gpu/embedding_grad_kernel.cu]: 同上，共性问题，麻烦都check一下

File: paddle/phi/kernels/sparse_weight_embedding_grad_kernel.h [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+#pragma once
+
+#include "paddle/phi/core/dense_tensor.h"
+#include "paddle/phi/core/selected_rows.h"
+
+namespace phi {
+
+template <typename T, typename Context>
+void SparseWeightEmbeddingGradKernel(const Context& ctx,
```
Comment (chenwhql) [inline paddle/phi/kernels/sparse_weight_embedding_grad_kernel.h:23]: SelectedRows相关的尽快移到selected_rows目录

File: paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc (hunk 15 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc]: 已修改

File: paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc (hunk 14 above)
Comment (phlrain) [reply paddle/phi/kernels/cpu/sparse_weight_embedding_grad_kernel.cc]: 已修改

**Review by chenwhql — APPROVED** [review]
如评论，selected rows kernel后续再调整下目录和形式

**Comment by phlrain** [comment]
> 如评论，selected rows kernel后续再调整下目录和形式

TODO已记录

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Comment by Liangwei-0521** [comment]
Traceback (most recent call last):
  File "D:\rank_v1.2\quant_predictor_2.py", line 191, in <module>
    results.extend(predictor.predict(batch_data, tokenizer))
  File "D:\rank_v1.2\quant_predictor_2.py", line 134, in predict
    self.predictor.run()
NotImplementedError: (Unimplemented) There are no kernels which are registered in the lookup_table_v2 operator.
  [Hint: Expected kernels_iter != all_op_kernels.end(), but received kernels_iter == all_op_kernels.end().] (at C:\home\workspace\Paddle\paddle\fluid\framework\operator.cc:1895)
  [operator < lookup_table_v2 > error
  
  embedding量化后，导入静态模型报错。

**Comment by Liangwei-0521** [comment]
embedding量化代码
def _quant_embeddings(input_prefix):
    
    # input_prefix: 静态模型路径
    # 最终量化的模型路径：
    output_filename_prefix = "rank_quant_emb"
    paddle.enable_static()
    place = paddle.set_device('cpu')
    exe = paddle.static.Executor(place)
    main_program, feed_target_names, fetch_targets = paddle.static.load_inference_model(input_prefix, exe)

    config = {
            "quantize_op_types": ["lookup_table_v2"], 
            "lookup_table_v2": {
                "quantize_type": "log",
                'quantize_bits': 8,
                'dtype': 'int8'
                }}

    quant_emb_program = quant.quant_embedding(main_program, place, config)

    input_dir = os.path.dirname(input_prefix)

    paddle.fluid.io.save_inference_model(
        input_dir,
        feed_target_names,
        fetch_targets,
        exe,
        quant_emb_program,
        model_filename=output_filename_prefix + ".pdmodel",
        params_filename=output_filename_prefix + ".pdiparams",
        export_for_deployment=True,
        program_only=False,
    )


if __name__ == '__main__':
    _quant_embeddings(input_prefix='./checkpoint/quant')

**Comment by zyfncg** [comment]
> Traceback (most recent call last): File "D:\rank_v1.2\quant_predictor_2.py", line 191, in results.extend(predictor.predict(batch_data, tokenizer)) File "D:\rank_v1.2\quant_predictor_2.py", line 134, in predict self.predictor.run() NotImplementedError: (Unimplemented) There are no kernels which are registered in the lookup_table_v2 operator. [Hint: Expected kernels_iter != all_op_kernels.end(), but received kernels_iter == all_op_kernels.end().] (at C:\home\workspace\Paddle\paddle\fluid\framework\operator.cc:1895) [operator < lookup_table_v2 > error
> 
> embedding量化后，导入静态模型报错。

@1998-Chen 这个问题在issue里提下吧，在PR下面不太好跟踪

**Pull request merged**